File: restoreIPAddresses.kt
class Solution {
    fun restoreIpAddresses(s: String): List<String> {
        val result = mutableListOf<String>()
        if (s.length < 4 || s.length > 12) return result

        fun backtrack(start: Int, dots: Int, current: String) {
            // If we've placed 3 dots and we're at the end of the string, it's valid
            if (dots == 3) {
                val part = s.substring(start)
                if (isValid(part)) {
                    result.add("$current$part")
                }
                return
            }

            // Try adding the dot after every 1, 2, or 3 digit substring
            for (i in start + 1 until s.length + 1) {
                val part = s.substring(start, i)
                if (isValid(part)) {
                    backtrack(i, dots + 1, "$current$part.")
                }
            }
        }

        // Helper function to validate a part
        fun isValid(part: String): Boolean {
            if (part.length > 3) return false
            if (part.length > 1 && part[0] == '0') return false
            return part.toIntOrNull()?.let { it in 0..255 } ?: false
        }

        // Start backtracking
        backtrack(0, 0, "")
        return result
    }
}

/*
fun main() {
    val solution = Solution()
    println(solution.restoreIpAddresses("25525511135"))
    println(solution.restoreIpAddresses("0000"))
    println(solution.restoreIpAddresses("101023"))
}
*/

File: restoreIPAddresses.go
package main

import (
	"fmt"
	"strconv"
)

func restoreIpAddresses(s string) []string {
	var result []string
	if len(s) < 4 || len(s) > 12 {
		return result
	}

	var backtrack func(start int, dots int, current string)
	backtrack = func(start int, dots int, current string) {
		// If we've placed 3 dots and we're at the end of the string, it's valid
		if dots == 3 {
			// Remaining part must be a valid number
			part := s[start:]
			if isValid(part) {
				result = append(result, current+"."+part)
			}
			return
		}

		// Try adding the dot after every 1, 2, or 3 digit substring
		for i := start + 1; i <= len(s) && i <= start+3; i++ {
			part := s[start:i]
			if isValid(part) {
				// Add the dot and recursively process the next part
				backtrack(i, dots+1, current+part+".")
			}
		}
	}

	// Initialize backtracking
	backtrack(0, 0, "")
	return result
}

func isValid(part string) bool {
	// If the part is longer than 3 digits, it's invalid
	if len(part) > 3 {
		return false
	}
	// If the part has leading zeros, it's invalid
	if len(part) > 1 && part[0] == '0' {
		return false
	}
	// Convert part to an integer and check if it's within the valid range
	num, err := strconv.Atoi(part)
	if err != nil {
		return false
	}
	return num >= 0 && num <= 255
}

/*
func main() {
	fmt.Println(restoreIpAddresses("25525511135"))
	fmt.Println(restoreIpAddresses("0000"))
	fmt.Println(restoreIpAddresses("101023"))
}
*/
